### PR TITLE
ipq806x: add support for Qxwlan E5200

### PR DESCRIPTION
**Pull request by zhangsd20**
Qxwlan E5200 based on IPQ8064

Specifications:
SOC:	Qualcomm IPQ8064
DRAM:	256 MiB
FLASH:	256 MiB Micron MT29F2G08ABBEAH4
SWITCH:	Qualcomm QCA8337
ETH:	5x 10/100/1000 Mbps Ethernet(RJ45)
INPUT:  1x Reset button(hw), 1x Reset button(sw), 1x wps button
LED:	1x Power ,2x usb ,1x esata, 1x wan ,4x lan, 2x wifi ,1x status(red/yellow)
SERIAL: UART (J13)
USB:	2x USB3.0
POWER:	1x DC jack for main power input (9-24 V)
SLOT:	2x Pcie (J1,J7)
ESATA:	1x esata

Flash instruction (using U-Boot CLI and tftp server):

 - Configure PC with static IP 192.168.1.10 and tftp server.
 - Rename "sysupgrade" filename to "firmware.bin" and place it in tftp
   server directory.
 - Connect PC with one of RJ45 ports, power up the board and press
   "enter" key to access U-Boot CLI.
 - Use the following command to update the device to OpenWrt: "run lfw".

Flash instruction (using U-Boot web-based recovery):

 - Configure PC with static IP 192.168.1.xxx(2-254)/24.
 - Connect PC with one of RJ45 ports, press the reset button, power up
   the board and keep button pressed for around 6-7 seconds, until LEDs
   start flashing.
 - Open your browser and enter 192.168.1.1, select "sysupgrade" image
   and click the upgrade button.

Signed-off-by: 张鹏 <sd20@qxwlan.com>

### CODE REVIEW

File: target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+*/
+
+#include "qcom-ipq8064-v1.0.dtsi"
+#include <dt-bindings/input/input.h>
+
+/ {
+	model = "Qxwlan E5200";
+	compatible = "qxwlan,e5200", "qcom,ipq8064";
+
+	memory@0 {
```
Comment (chunkeey) [inline target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: this should be memory@42000000 (the address value of the reg property)

File: target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+		ranges;
+		rsvd@41200000 {
+			reg = <0x41200000 0x300000>;
+			no-map;
+		};
+	};
+
+	aliases {
+		serial0 = &gsbi4_serial;
+		mdio-gpio0 = &mdio0;
```
Comment (chunkeey) [inline target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: Usually aliases are reserved for devices that the user interacts with (everything that is labeled, has a pictogram, etc...)

Also wouldn't it be great to wire up status_pass LED to led-boot, led-running? Any maybe status_fail to led-failsafe or use it as a panic-indicator?

File: target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+					       "gpio40", "gpio41", "gpio42",
+					       "gpio43", "gpio44", "gpio45",
+					       "gpio46", "gpio47";
+					function = "nand";
+					drive-strength = <10>;
+					bias-disable;
+				};
+				pullups {
+					pins = "gpio39";
+					bias-pull-up;
```
Comment (chunkeey) [inline target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: It's a bit odd that you set it to bias-disable in the "mux" node above. You should be able to remove the gpio39 from the list in "mux" above and add the function = "nand" + drive-strength to this node.

File: target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+				};
+				hold {
+					pins = "gpio40", "gpio41", "gpio42",
+					       "gpio43", "gpio44", "gpio45",
+					       "gpio46", "gpio47";
+					bias-bus-hold;
```
Comment (chunkeey) [inline target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: same as with "pullups". You could add the function and drive-strength to this node instead.

File: target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+					function = "gpio";
+					drive-strength = <8>;
+					bias-disable;
+				};
+			};
+
+			rgmii2_pins: rgmii2_pins {
+				mux {
+					pins = "gpio27", "gpio28", "gpio29", "gpio30", "gpio31", "gpio32",
+					       "gpio51", "gpio52", "gpio59", "gpio60", "gpio61", "gpio62" ;
```
Comment (chunkeey) [inline target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: Please split this up so that the lines are less than 81 characters long. As checkpatch will otherwise complain.

File: target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+				status = "ok";
+			};
+		};
+
+		sata-phy@1b400000 {
+			status = "ok";
+		};
+
+		sata@29000000 {
+			ports-implemented = <0x1>;
```
Comment (chunkeey) [inline target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: This change is upstream. Please make a separate patch which adds this line ```ports-implemented = <0x1>;``` to the qcom-ipq8064-v1.0.dtsi instead. (Or you could try to include "qcom-ipq8064.dtsi", where this was fixed by 064e4315750fbb80b21293c87cae0abb2c257d9b)

File: target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+					compatible = "qcom,smem";
+				};
+			};
+		};
+
+		mdio0: mdio {
+			compatible = "virtual,mdio-gpio";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			gpios = <&qcom_pinmux 1 0 &qcom_pinmux 0 0>;
```
Comment (chunkeey) [inline target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: Please make sure dt-bindings/gpio/gpio.h is included and reformat the phandle according to
https://www.kernel.org/doc/Documentation/devicetree/bindings/gpio/gpio.txt

File: target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+					0x0000c 0x80        /* PAD6_MODE */
+					0x000e4 0x6a545     /* MAC_POWER_SEL */
+					0x000e0 0xc74164de  /* SGMII_CTRL */
+					0x0007c 0x4e        /* PORT0_STATUS */
+					0x00094 0x4e        /* PORT6_STATUS */
+					>;
+			};
+
+			phy4: ethernet-phy@4 {
+				reg = <4>;
```
Comment (chunkeey) [inline target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: Just curious, is this the sfp-cage?

File: target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+		};
+
+		status_fail {
+			label = "e5200:red:status_fail";
+			gpios = <&qcom_pinmux 9 GPIO_ACTIVE_HIGH>;
+		};
+
+		esata {
+			label = "e5200:green:esata";
+			gpios = <&qcom_pinmux 26 GPIO_ACTIVE_HIGH>;
```
Comment (chunkeey) [inline target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: You could use the ```linux,default-trigger = "disk-activity";``` instead of the user space code in 01_leds.

File: target/linux/ipq806x/base-files/etc/board.d/01_leds [hunk 10]
```diff
@@ -33,6 +33,11 @@ netgear,r7800)
 	ucidef_set_led_netdev "wan" "WAN" "${boardname}:white:wan" "eth0"
 	ucidef_set_led_ide "esata" "eSATA" "${boardname}:white:esata"
 	;;
+qxwlan,e5200)
+	ucidef_set_led_usbport "usb1" "USB 1" "${boardname}:green:usb1" "usb1-port1" "usb2-port1"
```
Comment (chunkeey) [inline target/linux/ipq806x/base-files/etc/board.d/01_leds]: Could you please use the device-tree based usbport led trigger setup, I remember mentioning this in your previous
PR for your E2600AC C1/C2 https://github.com/openwrt/openwrt/pull/1655#discussion_r245761242 .

File: target/linux/ipq806x/base-files/etc/board.d/01_leds [hunk 11]
```diff
@@ -33,6 +33,11 @@ netgear,r7800)
...
+	ucidef_set_led_usbport "usb2" "USB 2" "${boardname}:green:usb3" "usb3-port1" "usb4-port1"
+	ucidef_set_led_ide "esata" "eSATA" "${boardname}:green:esata"
```
Comment (chunkeey) [inline target/linux/ipq806x/base-files/etc/board.d/01_leds]: (see below, this can be dropped with [disk-activity](https://github.com/openwrt/openwrt/pull/1935/commits/d4e5600c9c0583fd3a37d78350dd32c177605f81#r266672399) device-tree trigger ).

**Comment by chunkeey** [comment]
Can you please link to the product?

I tried to find it on: 
http://qxwlan.com/product.html


But the only board that comes close is the E5300AC:
http://qxwlan.com/product/p/96.html

**Comment by zhangsd20** [comment]
Hi,
The E5200 is not an E5300AC, I will quickly send my colleagues to our company website. 
Soon you will see.

File: target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts (hunk 4 above)
Comment (zhangsd20) [reply target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: Hi,
I saw that all models describe this node like this. How should I modify it?

File: target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts (hunk 3 above)
Comment (zhangsd20) [reply target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: I saw the qsdk and openwrt nodes seem to be described as such, so I also follow the same description.
If this is not correct, then how can I modify this node?
please help me,Thanks!

Comment (chunkeey) [reply target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: Well, in that case what does your contact at Qualcomm say about it? Either they must have a reason or they don't.

File: target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+					pins = "gpio54", "gpio65";
+					function = "gpio";
+					drive-strength = <2>;
+					bias-pull-up;
+				};
+			};
+
+			led_pins: led_pins {
+				mux {
+					pins = "gpio7", "gpio8", "gpio9","gpio26", "gpio53";
```
Comment (chunkeey) [inline target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: Please split this so the line length stays under 81.

File: target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+			status = "ok";
+		};
+
+		phy@110f8830 {
+			status = "ok";
+		};
+
+		usb30@0 {
+			status = "ok";
+
```
Comment (chunkeey) [inline target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: Extra empty line?

File: target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts (hunk 3 above)
Comment (zhangsd20) [reply target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: Sorry, I can't contact Qualcomm, I don't have their contact information.

**Comment by zhangsd20** [comment]
Hi,
This is E5200 device link.
http://www.qxwlan.com/product/p/105.html

**Comment by zhangsd20** [comment]
Hi,
   Please see what else needs to be modified.

**Comment by zhangsd20** [comment]
what should i do？

File: target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts (hunk 2 above)
Comment (ynezz) [reply target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: This should be addressed.

File: target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts (hunk 3 above)
Comment (ynezz) [reply target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: @Ansuel @pkgadd maybe you've an idea how to handle this?

**Comment by ynezz** [comment]
> what should i do?

* rebase it onto 4.19 kernel (and test it), provide full dmesg (file attachment) from running latest OpenWrt on the device
* address the remaining unresolved comments

File: target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts (hunk 3 above)
Comment (Ansuel) [reply target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: @ynezz I actually found out what @chunkeey  meant.
As he said it's strange to put bias-disable in the mux.... The best thing to do would be create a separate group with bias-disable. I'm converting every dts we have to follow this rule as this is what the kernel guy wants.

Comment (Ansuel) [reply target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: Here an example
https://github.com/Ansuel/openwrt/commit/a608ffe6e487d0466f245893cae4d9c0cde60d94

Comment (Ansuel) [reply target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-qxwlan-e5200.dts]: Here what @chunkeey suggested to do... Read some documentation and having a mux with no bias definition looks like it gets ignored by the driver
https://github.com/Ansuel/openwrt/commit/12b0bf3a5d9b72726f662d8573636afeade52e91 
This is how it should be defined... Tested on my r7800 and works like usual... (if it wasn't... then i wouldn't be able to boot AHAHAH)

**Comment by zhangsd20** [comment]
Ok, I will close this request and recreate a new one

**Comment by adrianschmutzler** [comment]
No, please use _this_ PR and do _not_ create a new one.

**Comment by zhangsd20** [comment]
Ok,
But this request has been for a long time, and this version kernel is 4.14 Without 4.19 kernel, how can I operate?

**Comment by pkgadd** [comment]
You can force-push your rebased patch set.

Disclaimer: I'm not an OpenWrt developer and can't help you with the merge itself, but a workable approach (according to my own opinion) would be:

The easiest (and imho most sensible approach) would be to fix up the remaining issues for kernel 4.14 (make sure not to accidentally commit the temporarily required/ local modifications to `target/linux/ipq806x/Makefile`), rebasing as necessary. Once you're happy with that, `git cp` the DTS from `target/linux/ipq806x/files-4.14/arch/arm/boot/dts/` to `target/linux/ipq806x/files-4.19/arch/arm/boot/dts/` (first commit) and adapt the DTS as necessary in a new (second) commit on top. Once you're happy, force push your branch and this pull request will be automatically updated with your changes.

As long as `target/linux/ipq806x/files-4.14/` and `target/linux/ipq806x/config-4.14` are still present in the master branch, this approach will retain the commit history and allow reviewing kernel 4.19 specific adaptions. If you (and the reviewers) are mostly happy with these changes and you're confident that it can be merged reasonably soon (before kernel 4.14 support for ipq806x gets completely removed from the master branch), this should work fine. If it will need more work or need longer than that to finalize, you'll may have to drop the kernel 4.14 additions in another rebase, but you can still keep working on this branch/ pull request.

**Comment by zhangsd20** [comment]
Thank you very much!
I have modified the issue on 4.14 kernel.
If there are no other issues, I will add my device on the 4.19 kernel and test it

File: target/linux/ipq806x/image/Makefile [hunk 14]
```diff
@@ -244,6 +244,18 @@ define Device/qcom_ipq8064-db149
...
 TARGET_DEVICES += qcom_ipq8064-db149
 
+define Device/qxwlan_e5200
+	$(call Device/LegacyImage)
+	$(call Device/UbiFit)
+	DEVICE_DTS := qcom-ipq8064-qxwlan-e5200
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	BOARD_NAME := e5200
+	DEVICE_TITLE :=  Qxwlan E5200
```
Comment (adrianschmutzler) [inline target/linux/ipq806x/image/Makefile]: DEVICE_TITLE has been replaced by DEVICE_VENDOR and DEVICE_MODEL.

File: target/linux/ipq806x/image/Makefile [hunk 15]
```diff
@@ -244,6 +244,18 @@ define Device/qcom_ipq8064-db149
 endef
 TARGET_DEVICES += qcom_ipq8064-db149
 
+define Device/qxwlan_e5200
+	$(call Device/LegacyImage)
+	$(call Device/UbiFit)
+	DEVICE_DTS := qcom-ipq8064-qxwlan-e5200
```
Comment (adrianschmutzler) [inline target/linux/ipq806x/image/Makefile]: Please have a look at recently merged commit e0382e88a919d342dd6fbf07971ceab15d0b6aaf.
Implications for you:
- Rename DTS files by removing "qxwlan-"
- Remove DEVICE_DTS here and add `SOC := qcom-ipq8064`
- Update 0069-arm-boot-add-dts-files.patch and DTS include lines.
- Same for to-be-added 4.19 changes.

**Comment by zhangsd20** [comment]
OK, 
I have modified and I tested my device on the latest master.
Wait until e5200 is supported on 4.14 kernel I will add 4.19 kernel support on the latest master branch

File: target/linux/ipq806x/image/Makefile (hunk 15 above)
Comment (zhangsd20) [reply target/linux/ipq806x/image/Makefile]: OK,
I have modified and I tested my device on the latest master.
Wait until e5200 is supported on 4.14 kernel I will add 4.19 kernel support on the latest master branch

File: target/linux/ipq806x/base-files/etc/board.d/02_network [hunk 16]
```diff
@@ -22,6 +22,10 @@ tplink,vr2600v)
 	ucidef_add_switch "switch0" \
 		"1:lan" "2:lan" "3:lan" "4:lan" "6@eth1" "5:wan" "0@eth0"
 	;;
+qxwlan,e5200)
+	ucidef_add_switch "switch0" \
+		"1:lan" "2:lan" "3:lan" "4:lan" "6u@eth1" "5:wan" "0u@eth0"
```
Comment (adrianschmutzler) [inline target/linux/ipq806x/base-files/etc/board.d/02_network]: Can you give details why this has to be untagged in the commit message?

Comment (zhangsd20) [reply target/linux/ipq806x/base-files/etc/board.d/02_network]: sorry,I don't know what you mean?
Is there a problem with my commit message?
 If there is a problem, it should be that I was negligent in the submission and made a mistake. I am sorry

Comment (Ansuel) [reply target/linux/ipq806x/base-files/etc/board.d/02_network]: Why 6u and 0u instead of tagged port

Comment (adrianschmutzler) [reply target/linux/ipq806x/base-files/etc/board.d/02_network]: As Ansuel said, you are using 6u and 0u, so connect these untagged while most other devices are tagged there. So, I assumed you have a reason for that and wanted you to describe that reason in the commit message. If you have no reason for that, please test with the tagged setup (6@eth1 and 0@eth0).

**Comment by Ansuel** [comment]
news on ths? @zhangsd20

**Comment by zhangsd20** [comment]
sorry,I don't know what you means

**Comment by Ansuel** [comment]
I mean any progress on this pr?

**Comment by zhangsd20** [comment]
I have modified it and tested it

File: target/linux/ipq806x/base-files/etc/board.d/02_network (hunk 16 above)
Comment (Ansuel) [reply target/linux/ipq806x/base-files/etc/board.d/02_network]: @zhangsd20 this still needs to be solved...

**Comment by Ansuel** [comment]
Also you need to rebase this pr and make rework the dts to support kernel 4.19

**Comment by zhangsd20** [comment]
Using 6 @ and 0 @ my device network ports are eth1.xx
and eth0.xx, this is not convenient for us to use; sometimes
we manually create a vlan is very inconvenient. So I use
6u @ and 0u @ here.

Doesn't this work?












在2020年01月15 08时46分，"Ansuel Smith"<notifications@github.com>写道：

@Ansuel commented on this pull request.
In target/linux/ipq806x/base-files/etc/board.d/02_network:
> @@ -22,6 +22,10 @@ tplink,vr2600v)
 	ucidef_add_switch "switch0" \
 		"1:lan" "2:lan" "3:lan" "4:lan" "6@eth1" "5:wan" "0@eth0"
 	;;
+qxwlan,e5200)
+	ucidef_add_switch "switch0" \
+		"1:lan" "2:lan" "3:lan" "4:lan" "6u@eth1" "5:wan" "0u@eth0"@zhangsd20 this still needs to be solved...

**Comment by Ansuel** [comment]
@zhangsd20 can't you just rebase all and force push ? I would be better to merge this in one go instead of 2 pr. 

Also about the tagged and untagged. I know it's not convenient but that's how openwrt works by default so pls just append this board to the default switch case.

```
case "$board" in
buffalo,wxr-2533dhp |\
compex,wpq864 |\
netgear,d7800 |\
netgear,r7500 |\
netgear,r7500v2 |\
qcom,ipq8064-ap148 |\
tplink,vr2600v |\
qxwlan,e5200)
	ucidef_add_switch "switch0" \
		"1:lan" "2:lan" "3:lan" "4:lan" "6@eth1" "5:wan" "0@eth0"
	;;
```

**Comment by zhangsd20** [comment]
ok, I have modified and submitted,
sorry, i don't konw just rebase all and force push.
I don't know what to do?

**Comment by Ansuel** [comment]
@zhangsd20 rebase this repo on top of master
https://stackoverflow.com/questions/51394172/git-rebase-branch-built-on-top-of-another-on-master

**Comment by adrianschmutzler** [comment]
> Using 6 @ and 0 @ my device network ports are eth1.xx
> and eth0.xx, this is not convenient for us to use; sometimes
> we manually create a vlan is very inconvenient. So I use
> 6u @ and 0u @ here.

If the device supports VLANs, it's actually the other way around, and it's convenient to already have the VLAN-based setup in place if you want to change it for some ports. Thus, using 6u and 0u is only the emergency solution when VLAN is not supported or broken. If it's working, please use 6@ and 0@ (note that this might also be the case why 0@ defaults to 0t@) as you have a switch with several ports here.

> Doesn't this work?

Just "it works" is not enough as prerequisite to get your code merged in OpenWrt trunk.

**Comment by adrianschmutzler** [comment]
> Because my branch has been submitted for a long time and there is no kernel4.19 in it. After this merge, I will resubmit support for 4.19kernel again.

Well, now we have kernel 4.19, so you will have to update. Since the merge has been only recently, it might help you to go back in git history and look at the changes there (there also was a set fixes a little later, both mostly by Ansuel and chunkeey). Those commits should reveal the necessary changes for your device.

IMO it should be enough to have kernel 4.19 support for the new device here, so no need to provide 4.14 as well unless you want to.

**Comment by Ansuel** [comment]
@zhangsd20 i merged your commit with the dts files update for 4.19 and 5.4 in the 5.4 kernel bump

**Comment by ynezz** [comment]
>  i merged your commit with the dts files update for 4.19 and 5.4 in the 5.4 kernel bump

@Ansuel Don't do that:

a) it makes no sense to waste review round-trip time with 4.14 kernel
b) adding new devices doesn't belong to the kernel bumps

Feel free to create separate PR for 4.19 and 5.4 for this device.

**Comment by Ansuel** [comment]
@ynezz ok then i will drop the commit from the 5.4 dump

**Comment by zhangsd20** [comment]
Hi,
    I delete my pr and resubmit it to the new kernel.












在2020年03月03 02时03分，"Petr Štetiar"<notifications@github.com>写道：

i merged your commit with the dts files update for 4.19 and 5.4 in the 5.4 kernel bump
@Ansuel Don't do that:
a) it makes no sense to waste review round-trip time with 4.14 kernel
b) adding new devices doesn't belong to the kernel bumps
Feel free to create separate PR for 4.19 and 5.4 for this device.

**Comment by adrianschmutzler** [comment]
> I delete my pr and resubmit it to the new kernel.

Just push the updated version to this branch, so the PR is kept.

**Comment by zhangsd20** [comment]
Hi,

   I tried how to push to a new branch last year, but I seem to have a problem with it,

   so I have never been successful. So I asked if I could merge my pr first, then I resubmit a new pr to the new kernel.


i don't konwn what should i do.













在2020年03月13 20时23分，"Adrian Schmutzler"<notifications@github.com>写道：


I delete my pr and resubmit it to the new kernel.
Just push the updated version to this branch, so the PR is kept.

**Comment by Ansuel** [comment]
@zhangsd20 do you use git from console or you are trying to push changes with the github webui?

If you are using git from console... just 
Save your changes in a separate directory

git reset --HARD origin/HEAD (to remove any changes in your local repo)
git pull (to update your repo)

And redo your changes

Then when you are finished... commit them and push them with --force

(git push --force)

**Comment by adrianschmutzler** [comment]
@zhangsd20 Please update this and use 5.4 kernel (only), as this is the current stable kernel for ipq806x.

**Comment by zhangsd20** [comment]
Hi,
I will delete the existing pr and recreate a pr












在2020年05月07 20时11分，"Adrian Schmutzler"<notifications@github.com>写道：


@zhangsd20 Please update this and use 5.4 kernel (only), as this is the current stable kernel for ipq806x.

**Comment by ynezz** [comment]
@zhangsd20 ping
